gneiss/sort.py

```diff
@@ -0,0 +1,81 @@
+import numpy as np
+import pandas as pd
+from functools import partial
+from gneiss.util import match
+
+
+def mean_niche_estimator(abundances, gradient):
+    """ Estimates the mean niche of an organism.
+
+    Calculates the mean niche or an organism along a gradient.
+    This is done by calculating the expected value of an organism
+    across the gradient.
+
+    Parameters
+    ----------
+    abundances : pd.Series, np.float
+        Vector of fraction abundances of an organism over
+        a list of samples.
+    gradient : pd.Series, np.float
+        Vector of numerical gradient values.
+
+    Returns
+    -------
+    np.float :
+        The mean gradient that the organism lives in.
+    """
+    if len(abundances) != len(gradient):
+        raise ValueError("Length of abundances (%d) doesn't match the length"
+                         " of the gradient (%d)" % (len(abundances),
+                                                    len(gradient)))
+    if np.any(pd.isnull(gradient)):
+        raise ValueError("`gradient` cannot have any missing values.")
+
+    # normalizes the proportions of the organism across all of the
+    # samples to add to 1.
+    v = abundances / abundances.sum()
+    m = np.dot(gradient, v)
+    return m
+
+
+def nichesort(table, gradient, niche_estimator=None):
+    """ Sort the table according to estimated niches.
+
+    Sorts the table by samples along the gradient
+    and otus by their estimated niche along the gradient.
+
+    Parameters
+    ----------
+    table : pd.DataFrame
+        Contingency table where samples are rows and
+        features (i.e. OTUs) are columns.
+    gradient : pd.Series
+        Vector of numerical gradient values.
+    niche_estimator : function
```

Comment:
add `, optional` and a description of the default (ok, to just add the name and add the function in the `See Also` numpy doc section).